File: Answer/3741.三个相等元素之间的最小距离-ii.ts
/*
 * @lc app=leetcode.cn id=3741 lang=typescript
 *
 * [3741] 三个相等元素之间的最小距离 II
 */

// @lc code=start
function minimumDistance(nums: number[]): number {
  const positions = new Map<number, [number] | [number, number]>();
  let best = Number.MAX_SAFE_INTEGER;

  nums.forEach((num, index) => {
    const state = positions.get(num);
    if (state === undefined) {
      positions.set(num, [index]);
      return;
    }

    if (state.length === 1) {
      positions.set(num, [state[0], index]);
      return;
    }

    best = Math.min(best, 2 * (index - state[0]));
    positions.set(num, [state[1], index]);
  });

  return best === Number.MAX_SAFE_INTEGER ? -1 : best;
}
// @lc code=end

File: Answer/3741.三个相等元素之间的最小距离-ii.cs
/*
 * @lc app=leetcode.cn id=3741 lang=csharp
 *
 * [3741] 三个相等元素之间的最小距离 II
 */

// @lc code=start
public class Solution
{
    public int MinimumDistance(int[] nums)
    {
        var positions = new Dictionary<int, (int older, int newer, int count)>();
        int best = int.MaxValue;
        foreach ((int num, int index) in nums.Select((num, index) => (num, index)))
        {
            if (!positions.TryGetValue(num, out var state))
            {
                positions[num] = (index, 0, 1);
                continue;
            }

            if (state.count == 1)
            {
                positions[num] = (state.older, index, 2);
                continue;
            }

            best = Math.Min(best, 2 * (index - state.older));
            positions[num] = (state.newer, index, 2);
        }

        return best == int.MaxValue ? -1 : best;
    }
}
// @lc code=end

